service/instance.py

```diff
@@ -1315,12 +1317,31 @@ def _test_for_licensing(esh_machine, identity):
 
 
 def check_allocation(username, allocation_source):
+    logger.debug('check_allocation - username: %s', username)
+    logger.debug('check_allocation - allocation_source: %s', allocation_source)
+    if not settings.ENFORCING:
+        logger.debug('check_allocation - settings.ENFORCING is False - just returning')
+        return
     user = AtmosphereUser.objects.filter(username=username).first()
     if not user:
         raise Exception("Username %s does not exist" % username)
     compute_remaining = allocation_source.time_remaining(user)
     over_allocation = compute_remaining < 0
-    if over_allocation and settings.ENFORCING:
+
+    enforcement_override_choice = AllocationSourcePluginManager.get_enforcement_override(user,
+                                                                                         allocation_source)
+    logger.debug('check_allocation - enforcement_override_choice: %s', enforcement_override_choice)
+    if over_allocation and enforcement_override_choice == EnforcementOverrideChoice.NEVER_ENFORCE:
```

Comment:
You can simplify the logic here for a readability win. The user's compute remaining doesn't actually matter (and could be computed after this if check).
```
    if enforcement_override_choice == EnforcementOverrideChoice.NEVER_ENFORCE:
        return
    elif enforcement_override_choice == EnforcementOverrideChoice.ALWAYS_ENFORCE:
        # Using a dummy value of -1, should throw a separate error
        raise OverAllocationError(allocation_source.name, -1)
```

Reply:
I did consider using simpler logic like you have, but decided against it. I chose to use this more explicit and verbose logic with the matching logging statements to help some poor future programmer to figure out how this feature is supposed to work.

Reply:
The current logic was unclear to me, and I think there's more to it than a diff of opinion. 

This code has 1 of 2 behaviors and the fresh reader is going to try and understand how ALWAYS/NEVER work. Either ALWAYS/NEVER are absolute, or they only come into effect when the user is over. If they see that over allocation is part of the function, they should assume that ALWAYS/NEVER play a role when as is over_allocation. However, the actual behavior is that these checks don't take over_allocation into account at all.

Reply:
Hmm... Fair enough. I see your point. I'll modify it.

Reply:
@cdosborn Change made.